### ISSUE TITLE
Data race in init_simd() - all platforms

### ISSUE THREAD

**Issue by hasinoff** [libjpeg-turbo/libjpeg-turbo#87]
As acknowledged in comments, init_simd() is racy in multi-threaded environments, with a potential data race in initializing 'simd_support' variable. The same race is present for all platforms (simd/jsimd_arm64.c, simd/jsimd_x86_64.c, etc).

Although the race looks 'benign', it would be nice to clean this up to avoid noise from tsan or related tools. The only issue I can see is if the environment variables JSIMD_FORCENONE or JSIMD_NOHUFFENC were changed in the middle of the data race, and even then the consequences seem mild - possibly unintended debugging mode for one of the threads.

In our environment, this race shows up in our tsan-instrumented fuzz tests that use libjpeg-turbo. We've worked around the issue by adding _Thread_local to the relevant globals ('simd_support' etc.) but this is not a general solution - not all toolchains are C11 compliant.

**Comment by dcommander** [comment]
I agree that it would be nice, but I don't know how to do it without introducing unwanted dependencies, such as a dependency on pthreads or on C11.  Linaro attempted to fix this by calling init_simd() when the .so is dynamically loaded, but of course that does no good if libjpeg-turbo is linked statically. As you point out, this is benign, and fixing it is not a priority.  I'm happy to entertain solutions that fix it, provided that they don't introduce unwanted complications, but I'm not going to spend any of my personal time looking at it.

**Comment by mayeut** [comment]
If there's no requirement for the solution to work on all platforms (given that the race looks benign) then something simple can probably be achieved using ```__sync_val_compare_and_swap``` & ```__GCC_HAVE_SYNC_COMPARE_AND_SWAP_4``` for clang/gcc or ```InterlockedCompareExchange``` when under MSVC.

With such a solution, only the first thread that "successfully" called ```__sync_val_compare_and_swap``` will set the global variable that all other threads would use.  
Would that be something that could be considered ?

**Comment by dcommander** [comment]
Possibly.  `InterlockedCompareExchange()` is part of kernel32.dll, so it wouldn't introduce any new dependencies on Windows, but I'd need to know which versions of MSVC support it.  Similarly for GCC-- what is the earliest version that supports this feature?

**Comment by dcommander** [comment]
Also, someone else would have to author a pull request.  I have zero time to look into this.

**Comment by ssvb** [comment]
If I understand it correctly, the worst thing that can happen is that the runtime detection code runs twice. And if somebody is changing the environment variables at the wrong time, then this is racy regardless of whether the libjpeg-turbo has any safeguard against running the runtime detection code twice or not. The only potential concern is whether 32-bit reads/writes are really atomic on every supported platform or not.

Also it looks like @hasinoff mostly wants to shut up some automatic analyser :) So any solution that is still "unfriendly" to this particular automatic analyser would not make things better.

> We've worked around the issue by adding _Thread_local to the relevant globals ('simd_support' etc.) but this is not a general solution - not all toolchains are C11 compliant.

But this can be detected by the build scripts and only enabled when the toolchain supports this feature, right? It seems to be an alright solution to me if the only practical goal is compatibility with data race detection tools.

**Comment by dcommander** [comment]
The set of people who might want to run libjpeg-turbo through tsan is probably greater than the set of people who build libjpeg-turbo using C11-compatible toolchains.  Ideally, this would be something that is built into the "official" binaries, which means it will be thoroughly tested.  I don't like a lot of #ifdefs hanging around, particularly when I don't personally build the code within the #ifdef.

**Comment by ssvb** [comment]
Maybe I'm missing something, but [tsan](http://clang.llvm.org/docs/ThreadSanitizer.html) seems to be a Clang specific tool. I believe that Clang supports the __thread keyword since a long time ago (even without the strict C11 mode), and so does GCC.

**Comment by ssvb** [comment]
There is also the "__attribute__((no_sanitize_thread))" tweak mentioned in the tsan readme, most likely specifically for having suppressions for the known benign cases.

**Comment by dcommander** [comment]
I am hesitant to use `__thread`, because on Linux, it is implemented using opaque mutexes.  I've had problems with that in the past (not in libjpeg-turbo but in other projects I develop.)  I would be OK with the attribute() approach, if it was guarded by an appropriate clang-specific #ifdef.

**Comment by mayeut** [comment]
Something like the following should work (not tested though):
```
#ifdef __has_attribute
# if __has_attribute(no_sanitize)
#   define LJT_NOSANITIZE(kind) __attribute__((no_sanitize(kind)))
# endif
#endif
#ifndef LJT_NOSANITIZE
# define LJT_NOSANITIZE(kind)
#endif
```

```
LOCAL(void)
init_simd (void) LJT_NOSANITIZE("thread")
{
...
```

**Comment by dcommander** [comment]
@hasinoff will have to confirm whether that proposed patch eliminates the false positives in tsan, but if it does, I'm OK with including it.

**Comment by hasinoff** [comment]
I tested the LJT_NOSANITIZE code above from @mayeut and it's effective at suppressing tsan's complaints about the data race.

**Comment by dcommander** [comment]
Unfortunately the code above does not eliminate the warnings with the GCC 6 implementation of tsan, which doesn't seem to support the no_sanitize attribute.  It does support the no_sanitize_thread attribute, but that also doesn't eliminate the warnings.  For my money, if this doesn't generally eliminate warnings with all versions of tsan, it isn't worth checking in.

I also spent significant time investigating how this issue might be fixed properly.  It is certainly possible to store the simd_support and simd_huffman variables per-object, by using the opaque master structs.  However, this would mean that the CPU detection code would run every single time a libjpeg struct is created, and since that CPU detection code requires reading /proc on some platforms, it's not something we want to run any more often than necessary.  IMHO, it's less of a sin for the code to be racy.  Worst case with the current code is that one or two of the threads miss the update from another thread, and the CPU detection runs two or three times during the execution of the program.  Making simd_support a per-object variable would mean that the CPU detection code might run hundreds or even thousands of times during the execution of a program (think about browsers-- jsimd_simd_cpu_support() would run every single time an image is decoded.  Yuck.)  I don't know of another way to fix this properly without using TLS, which is generally going to introduce unwanted dependencies.

Depending on how a particular multi-threaded program is structured, it might be possible to eliminate data races in some of those programs by guaranteeing that init_simd() executes within the body of jpeg_CreateCompress() or jpeg_CreateDecompress().  That would work if the multi-threaded program creates a per-thread libjpeg object prior to launching the threads, but I predict that it wouldn't fix the issue for all programs.  Apart from the constructor functions, everything else in the API is per-object.

**Comment by davidben** [comment]
Perhaps the pthread/Windows dependency could be gated on a build flag? And then it could call `pthread_once` on POSIX and `InitOnceExecuteOnce` on Windows.

**Comment by dcommander** [comment]
> Perhaps the pthread/Windows dependency could be gated on a build flag? And then it could call `pthread_once` on POSIX and `InitOnceExecuteOnce` on Windows.

The race condition is completely innocuous, so it isn’t worth introducing those dependencies.

**Comment by davidben** [comment]
It's only _unlikely_ to cause problems. No race condition is _completely_ innocuous. Programs with [data races](https://en.cppreference.com/w/c/language/memory_model) are [undefined](https://en.cppreference.com/w/c/language/behavior) in C and C++. In principle, the compiler has no obligations on the behavior of problems that exhibit undefined behavior. The program may be optimized in a way that makes it no longer innocuous.

**Comment by dcommander** [comment]
@davidben To the best of my understanding, the `simd_support` variable is atomic, so the worst case is that two threads might try to initialize the SIMD extensions simultaneously, and the second thread might waste effort by determining the correct value of `simd_support` while the first thread is doing likewise.  The code has been that way since 2010, and no user-visible problems have been reported from it despite it being used by billions of people every day.

That being said, it would be straightforward to make the `simd_support` variable thread-local on platforms that support that, since the libjpeg-turbo build system already detects support for TLS.  Historically, there were some platforms that didn't support TLS, but to the best of my knowledge, we no longer officially support any of those platforms.

**Comment by dcommander** [comment]
I made all of the `simd_*` variables thread-local, which seems to appease TSan.  Let me know if you encounter any issues.

**Comment by dcommander** [comment]
The SIMD dispatchers have been completely overhauled in the dev/3.2 Evolving branch.  They are now initialized on a per-instance basis rather than a per-thread basis, thus eliminating the dependence on TLS.
